source/agora/common/Hash.d

```diff
@@ -289,3 +289,182 @@ nothrow @nogc @safe unittest
+
+*******************************************************************************/
+
+public struct PreImageCache
+{
+    @safe nothrow:
+
+    /// Store the actual preimages
+    private Hash[] data;
+
+    /// Interval between two preimages in `data`
+    private const ulong interval;
+
+    /// Construct an instance using already existing data
+    public this (inout(Hash)[] data_, ulong sample_size) inout @nogc pure
+    {
+        this.data = data_;
+        this.interval = sample_size;
+    }
+
+    /***************************************************************************
+
+        Construct an instance and allocate memory
+
+        This takes a `count` of pre-image and a sample size, or distance.
+        For example, if one wish to store 1000 pre-images in this cache,
+        and perform no more than 5 hash operations each time,
+        the `sample_size` should be `5` and the `count` should be `200`
+        (1000 / 5).
+
+        Params:
+          count = Number of samples to store (number of entries in the array)
+          sample_size = Distance between two pre-images
+
+    ***************************************************************************/
+
+    public this (ulong count, ulong sample_size) pure
+    {
+        assert(count > 1, "A count of less than 2 does not make sense");
+        assert(sample_size > 1, "A distance of less than 2 does not make sense");
+
+        this.interval = sample_size;
+        this.data = new Hash[](count);
+    }
+
+    /***************************************************************************
+
+        Populate this cache from the `seed`
+
+        Params:
+          seed = Initial value to derive preimage from. This will always be the
+                 first value of the array.
+          length = The number of entries to populate.
+                   This can be used when large sample size are used,
+                   and one wish to stop initialization past a certain threshold.
+                   The special value `0` means "everything", and is the default.
+
+        Returns:
+          The result of hashing seed `sample_size * (count + 1) - 1` times
+          if `length == 0`, or `sample_size * (length + 1) - 1` otherwise,
+          provided `length <= count`.
+
+    ***************************************************************************/
+
+    public Hash reset (Hash seed, size_t length = 0)
+    {
+        // Set the length, so that extra entries are not accessible through
+        // `opIndex`
+        if (length)
+        {
+            this.data.length = length;
+            () @trusted { assumeSafeAppend(this.data); }();
+        }
+        size_t count = 0;
+        immutable end = this.data.length * this.interval - 1;
+        foreach (idx, ref entry; this.data)
+        {
+            entry = seed;
+            do
+                seed = hashFull(seed);
+            while (++count % end % this.interval);
```

Comment:
The association is always left to right in D, correct? `A % B % C` => `(A % B) % C`?

You can leave it as-is if it should be obvious.

Reply:
Correct